content/en/blog/_posts/2021-09-20-admission-controllers-and-container-drift/index.md

```diff
@@ -0,0 +1,80 @@
+
+### 1.  Receive the admission request of interactive kubectl command
+We need to first define qualified admission webhooks to receive specific requests in our admission controller service. In the interest of `kubectl exec`, for example, we configure the webhook’s rules with resources being “pods/exec” and operations as “CONNECT”. They tell the K8s API server that all “exec” requests should be subject to our admission controller service. We also specify a “url” which gives the location of our service and “caBundle” to provide its certificate, both under the “clientConfig” stanza. Here is a short example of what our ValidatingWebhookConfiguration object looks like:
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: "example-webhook"
+webhooks:
+- name: "admit-pod-exec-webhook"
+  rules:
+  - apiGroups: ["*"]
+    apiVersions: ["*"]
+    operations: ["CONNECT"]
+    resources: ["pods/exec"]
+  clientConfig:
+    url: "https://example:9443/admit-pod-exec"
+    caBundle: "<PEM encoded CA bundle>"
+```
+
+### 2.  Identify the pod with potentially mutated container
+Once a request of `kubectl exec` comes to our admission controller service, it identifies the target pod by adding a customized K8s label. In this way, we can not only query all the pods with potentially mutated containers, but also enable the service to retrieve previously identified pods in case it gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object given in the admission request is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the label by a Kube client connecting to the cluster. The service always returns an admission response as allowed after passing that pod reference to its Kube client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected simply by running a `kubectl describe` command:
```

Comment:
```suggestion
Once a request of `kubectl exec` comes to our admission controller service, it identifies the target pod by adding a customized K8s label. In this way, we can not only query all the pods with potentially mutated containers, but also enable the service to retrieve previously identified pods in case it gets restarted. The service cannot directly label the pod and return the updated object in its admission response. This is because the object given in the admission request is of type “PodExecOptions”, a child object of type “Pod”. As a result, there is a separate process in our admission controller service that patches the label by a Kube client connecting to the cluster. The service always returns an admission response as allowed after passing that pod reference to its Kube client, who is responsible for labeling and sending related K8s events to the target pod. Developers can check whether their pods are affected or not, simply by running a `kubectl describe` command:
```
A general suggestion you might want to include.